src/sentry/dynamic_sampling/tasks/task_context.py

```diff
@@ -84,4 +89,104 @@ def to_dict(self) -> Dict[str, Any]:
         }
         if self.context_data is not None:
             ret_val["taskData"] = {k: v.to_dict() for k, v in self.context_data.items()}
+            # update the execution time for each function
+            for name, state in self.context_data.items():
+                state.execution_time = self.timers.current(name)
         return ret_val
+
+
+@dataclass
+class TimerState:
+    elapsed: float
+    started: Optional[float]
+
+
+class Timers:
+    """
+    Keeps the state of many named timers.
+
+    Individual timers will be created at first use.
+    Users should not use this class directly, but use the NamedTimer class instead.
+    TaskContext uses this class to keep track of timers.
+    For using timers for functions use TaskContext.get_timer("<name>") to get a NamedTimer.
+
+    """
+
+    def __init__(self):
+        self.timers: Dict[str, TimerState] = {}
+
+    def get_timer(self, name: str) -> "NamedTimer":
+        return NamedTimer(name, self)
```

Comment:
I personally dislike the control flow of such implementation because it's quite cumbersome, however, from the usage end it's quite nice.

Another possibility which has an easier control flow would be to have the timers class and you can register timers on it and each timer has a state on its own, then via the name you get the timer and act upon it. This will still give you the semantics that you implemented here, that is, you will be able to access the same instance in subsequent calls.

```python
timers = Timers()
	.register(NamedTimer("name_1"))
	.register(NamedTimer("name_2"))
		
timers.get_timer("name_1").start()
```

No need to change your implementation, just thinking out loud my ideas.

Reply:
Don't quite get why you want another step. 
There is no need for `register`.
Just ask timers for a named timer and you will get one.
Actually timers is now also hidden ( you don't really need to worry about it), you just ask the `TaskContext` for a timer and it takes care of everything. 
The idea is that you pass around the `TaskContext` which handles state & timers. 
When you want a timer you ask for it and it will give you the timer with the requested name.